File: app/src/main/java/com/weathercock/cleanarchitecture/presentation/presenter/fragment/AboutFragmentViewModel.java
package com.weathercock.cleanarchitecture.presentation.presenter.fragment;

import android.content.Context;
import android.content.Intent;
import android.databinding.ObservableField;
import android.net.Uri;
import android.view.View;

import com.weathercock.cleanarchitecture.CleanApplication;
import com.weathercock.cleanarchitecture.R;
import com.weathercock.cleanarchitecture.domain.usecase.AboutUseCase;

import javax.inject.Inject;

/**
 * About Fragment ViewModel
 *
 * @author Kensuke
 */
public class AboutFragmentViewModel {

    public ObservableField<String> verText;
    public ObservableField<String> developByText;
    public ObservableField<String> copyrightText;

    private Context context;

    @Inject
    public AboutUseCase aboutUseCase;

    public AboutFragmentViewModel(Context context) {
        CleanApplication.getInstance(context).getApplicationComponent().inject(this);
        this.context = context;
        this.verText = new ObservableField<>(context.getString(R.string.about_ver, aboutUseCase.getCurrentVersion()));
        this.developByText = new ObservableField<>(context.getString(R.string.about_develop_by, context.getString(R.string.developer_name)));
        this.copyrightText = new ObservableField<>(context.getString(R.string.about_copyright, aboutUseCase.getCurrentYear(), context.getString(R.string.developer_name)));
    }

    public void onClickPlayStore(View view) {
        toPlayStore();
    }

    public void onClickMail(View view) {
        toMailApp();
    }

    public void onClickWebSite(View view) {
        toWebSite();
    }

    private void toPlayStore() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(aboutUseCase.getPlayStoreUrl()));
        context.startActivity(intent);
    }

    private void toMailApp() {
        Intent intent = new Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:" + aboutUseCase.getMailUrl()));
        context.startActivity(intent);
    }

    private void toWebSite() {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(aboutUseCase.getWebSiteUrl()));
        context.startActivity(intent);
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/presentation/presenter/fragment/AboutFragmentViewModel.kt
package com.kazakago.cleanarchitecture.presentation.presenter.fragment

import android.content.Context
import android.content.Intent
import android.databinding.ObservableField
import android.net.Uri
import android.os.Bundle
import android.view.View
import com.github.salomonbrys.kodein.LazyKodein
import com.github.salomonbrys.kodein.LazyKodeinAware
import com.github.salomonbrys.kodein.android.appKodein
import com.github.salomonbrys.kodein.instance
import com.kazakago.cleanarchitecture.R
import com.kazakago.cleanarchitecture.domain.usecase.appInfo.GetAppVersionUseCase
import com.kazakago.cleanarchitecture.domain.usecase.appInfo.GetMailAddressUrlUseCase
import com.kazakago.cleanarchitecture.domain.usecase.appInfo.GetOfficialSiteUrlUseCase
import com.kazakago.cleanarchitecture.domain.usecase.appInfo.GetPlayStoreUrlUseCase
import com.kazakago.cleanarchitecture.presentation.listener.presenter.fragment.AboutFragmentViewModelListener
import java.util.*

class AboutFragmentViewModel(private val context: Context, private val listener: AboutFragmentViewModelListener): LazyKodeinAware {

    override val kodein = LazyKodein(context.appKodein)

    val verText = ObservableField<String>()
    val developByText = ObservableField<String>()
    val copyrightText = ObservableField<String>()

    private val getAppVersionUseCase: GetAppVersionUseCase by instance()
    private val getPlayStoreUrlUseCase: GetPlayStoreUrlUseCase by instance()
    private val getMailAddressUrlUseCase: GetMailAddressUrlUseCase by instance()
    private val getOfficialSiteUrlUseCase: GetOfficialSiteUrlUseCase by instance()

    fun onViewCreated(savedInstanceState: Bundle?){
        verText.set(context.getString(R.string.about_ver, getAppVersionUseCase.execute(Unit)))
        developByText.set(context.getString(R.string.about_develop_by, context.getString(R.string.developer_name)))
        copyrightText.set(context.getString(R.string.about_copyright, Calendar.getInstance().get(Calendar.YEAR), context.getString(R.string.developer_name)))
    }

    fun onClickPlayStore(view: View?) {
        toPlayStore()
    }

    fun onClickMail(view: View?) {
        toMailApp()
    }

    fun onClickWebSite(view: View?) {
        toWebSite()
    }

    private fun toPlayStore() {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(getPlayStoreUrlUseCase.execute(Unit)))
        listener.startActivity(intent = intent)
    }

    private fun toMailApp() {
        val intent = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:" + getMailAddressUrlUseCase.execute(Unit)))
        listener.startActivity(intent = intent)
    }

    private fun toWebSite() {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(getOfficialSiteUrlUseCase.execute(Unit)))
        listener.startActivity(intent = intent)
    }

}
